lib/rules/disallow-trailing-whitespace.js

```diff
@@ -64,13 +64,70 @@ module.exports.prototype = {
 
     check: function(file, errors) {
         var ignoreEmptyLines = this._ignoreEmptyLines;
-
         var lines = file.getLines();
+
         for (var i = 0, l = lines.length; i < l; i++) {
+
             if (lines[i].match(/\s$/) && !(ignoreEmptyLines && lines[i].match(/^\s*$/))) {
-                errors.add('Illegal trailing whitespace', i + 1, lines[i].length);
+                var fixed = false;
+                var currentLineNumber = i + 1;
+                var startLineNumber;
+                var precendingToken;
+                var targetToken;
+
+                while (!precendingToken && currentLineNumber > 0) {
+                    precendingToken = file.getLastTokenOnLine(currentLineNumber, {includeComments: true});
+                    currentLineNumber--;
+                }
+
+                if (typeof precendingToken === 'undefined') {
+                    targetToken = file.getFirstToken();
+                    startLineNumber = 1;
+                } else {
+                    targetToken = file.getNextToken(precendingToken, {includeComments: true});
+                    startLineNumber = precendingToken.loc.end.line;
+
+                    if (precendingToken.isComment &&
+                        precendingToken.loc.start.line <= (i + 1) &&
+                        precendingToken.loc.end.line >= (i + 1)) {
+
+                        if (precendingToken.type === 'Block') {
+
+                            if (ignoreEmptyLines) {
+                                var blockLines = precendingToken.value.split(/\n/);
+
+                                for (var k = 0; k < blockLines.length; k++) {
+
+                                    if (!blockLines[k].match(/^\s*$/) || k === 0) {
+                                        blockLines[k] = blockLines[k].split(/\s$/).join('');
+                                    }
+                                }
+                                precendingToken.value = blockLines.join('\n');
+                            } else {
+                                precendingToken.value = precendingToken.value.split(/\s\n/).join('\n');
+                            }
+                        } else {
+                            precendingToken.value = precendingToken.value.split(/\s$/).join('');
+                        }
+                        fixed = true;
+                    }
+                }
+
+                if (typeof targetToken !== 'undefined' && !fixed) {
+                    var eolCount = targetToken.loc.start.line - startLineNumber + 1;
+                    var targetIndent = '';
+                    var targetLine = lines[targetToken.loc.start.line - 1];
+                    for (var j = 0, whitespace = targetLine.charAt(j);
```

Comment:
this can be rewritten as `for (var j = 0, whitespace ; whitespace = targetLine.charAt(j), whitespace.match(/\s/); j++)`